### PR TITLE
Feat: Icon HTTP Server

### PR DESCRIPTION
**Pull request by tomasciccola**
This would solve #292

### CODE REVIEW

File: src/mapeo-project.js
Comment (achou11) [inline src/mapeo-project.js:216]: are we creating an icon server for each project? my initial thought is that this would be handled similarly to how we're planning to do blobs i.e. one shared server with routes prefixed by the kind of server (e.g. `/blob/` or `/icon/`) and the routes accepting a `projectId` param to work across all projects

File: src/icon-server/fastify-plugin.js
Comment (achou11) [inline src/icon-server/fastify-plugin.js:41]: should this be `':projectId/:iconDocId/:size/:pixelDensity'`? my understanding is that the plan is to have a single server instance that will host these routes.

**Review by achou11 — CHANGES_REQUESTED** [review]
have some questions about the end-usage of this plugin:

- my understanding is that the plugin should work across several projects, similar to the blob plugin.

- in terms of design, my thinking is that the plugin should be able to be used like so (in our case this would probably happen in the MapeoManager implementation):

```ts
const fastify = fastify()

fastify.register(BlobServerPlugin, { prefix: '/blob/', ... })
fastify.register(IconServerPlugin, { prefix: '/icon/', ... })
```

File: src/icon-server/fastify-plugin.js
Comment (tomasciccola) [reply src/icon-server/fastify-plugin.js:73]: yeah yeah. Most of it is incomplete, this is just a stub so that we please ts. I didn't realize I made a mistake and didn't make it a draft.

File: src/icon-server/fastify-plugin.js
Comment (achou11) [reply src/icon-server/fastify-plugin.js:50]: scratch that: thinking that maybe getByDocId implementation should be updated to throw when no match is found, which means you'd be able to keep it like this

Comment (tomasciccola) [reply src/icon-server/fastify-plugin.js:50]: heh, nice catch. `core.get`  can also return undefined. So I should handle both cases and return a 404

File: src/mapeo-project.js
Comment (tomasciccola) [reply src/mapeo-project.js:216]: This first commit just starts to put everything into place by using the blob-server implementation as a reference. 
For now the blob-server is instantiated inside `mapeo-project` so I was assuming it was handled per project. 
But what you say - I think - makes sense: only have one server per MapeoManager and use `:projectId` as a param.

Comment (achou11) [reply src/mapeo-project.js:216]: > For now the blob-server is instantiated inside mapeo-project so I was assuming it was handled per project.

the `BlobApi` is instantiated in each project, but the server isn't. We have yet to actually integrate any server into mapeo manager/project

Comment (tomasciccola) [reply src/mapeo-project.js:216]: Yeah, I think #291 is made to implement that

Comment (tomasciccola) [reply src/mapeo-project.js:216]: mhm, from what I see in `mapeo-project.js` in main, it seems that the server is instantiated on every project (see [here](https://github.com/digidem/mapeo-core-next/blob/9d465d388e76633fe54c0df46d999cb6142f4948/src/mapeo-project.js#L202)). But maybe I'm missing smth and `createBlobServer` acts as a singleton that instantiates the server if it doesn't exists yet??

Comment (achou11) [reply src/mapeo-project.js:216]: ah no, you're right. totally missed that and was only fixated on the blob api instantiation, my bad 😅

Comment (tomasciccola) [reply src/mapeo-project.js:216]: > ah no, you're right. totally missed that and was only fixated on the blob api instantiation, my bad 😅

I mean, it makes sense to have only one server per `MapeoManager` and not per `MapeoProject`. More so if the `projectId` is a param to the server; if not, it kinda feels redundant to pass the `projectId` ? (since we're always talking about the same instance of `MapeoProject` ).

**Comment by achou11** [comment]
my thinking is that each of the Fastify plugins (so both blobs and icons) should accept an option that is either:

1. The manager instance (like you have it so far)
2. An async function called `getProject()`.

The fundamental thing that each plugin eventually needs is to access project-specific resources, so we need to define a way for the plugin to access a project and its API. For this reason, I _think_ 2 makes a little more sense than 1, but either should work.

In my opinion, the end goal is to be able to set up the server in `MapeoManager` like so:

```ts
this.#fastifyServer = new Fastify({...})

// Register the blob routes
this.#fastifyServer.register(BlobServerPlugin, {
  prefix: '/blob/',
  // One of the following...
  // 1. The manager option
  manager: this,
  
  // 2. The getProject option
  getProject: async (projectId) => {
   return this.#getProject(projectId)
  }
)

// Register the icon routes
this.#fastifyServer.register(IconServerPlugin, {
  prefix: '/icon/',
  // One of the following...
  // 1. The manager option
  manager: this,
  
  // 2. The getProject option
  getProject: async (projectId) => {
   return this.#getProject(projectId)
  }
)
```

At a glance, a shortcoming with this approach is that you may need to expose random instance properties that aren't necessarily useful to a frontend client (remember, we're going to proxy this when interfacing with an IPC layer). I _think_ a  solution to that is using the symbol key accessor pattern that we're using to expose certain methods/fields but for internal usage + testing. For example, if you need the core manager instance, you can set it up like so:

```ts
export const kCoreManager = Symbol('coreManager')

class MapeoProject {
  get [kCoreManager]() {
    return this.#coreManager
  }
}

// Somewhere else where you have the project instance:
import {kCoreManager} from '...' // must explicitly import and use this symbol if you want to access that exposed getter
const coreManager = project[kCoreManager]
```

File: src/icon-server/fastify-plugin.js
Comment (tomasciccola) [inline src/icon-server/fastify-plugin.js:13]: I commented this regexes 'cause I was getting validation errors and I couldn't bother with regex today.
It seems that the project key is actually 26 bytes? (printing `projectId` that's returned from `manager.createProject()` yields `52`...)
Anyways, there's probably something I'm missing...

Comment (achou11) [reply src/icon-server/fastify-plugin.js:13]: that number seems familiar for some reason, but the project id that's used for the url _should_ be 32 bytes since we're using this from @mapeo/crypto to create them:

https://github.com/digidem/mapeo-crypto/blob/a3ec70b97318260e2f1593d6abbaeda065cc9407/utils.js#L39

Comment (achou11) [reply src/icon-server/fastify-plugin.js:13]: ah yeah the encoding is different, which produces an id of 26 bytes. The projectId here is really the project's "public" id, which is a one-way hashed version of the project's "internal" id. Think it's a matter of updating this regex to assume length of 52 characters

Needs to be done for the blob server plugin too (but that's out of scope here)

Comment (tomasciccola) [reply src/icon-server/fastify-plugin.js:13]: Mhm, following the trail, the `keys.projectKey` on `mapeo-manager:217` seems to be 32-bytes long. Now, after passing it to `projectKeyToPublicId` (which wraps `keyToPublicId`) the resulting string length is 52 (shouldn't it be 64?).
I see in that fn that you linked that it uses some encoding called z-base-64. I'm not familiar with that, but does that changes something??
I see that I'm on mapeo-crypto-alpha.8 which is the last version, but at the same time on my local version of `mapeo-crypto` that function is named `projectKeyToPublicId` instead of `keyToPublicId` (I've `rm -rf node_modules && npm install` just in case...)

Comment (achou11) [reply src/icon-server/fastify-plugin.js:13]: @tomasciccola yeah not sure if you saw my last comment but the project id we're returning from `createProject` is not a hex-encoded string but rather the z-base-64 one. this is expected, i just forgot that it's a different length and the only change needed here is updating the regex to use 52 instead of 64, since this API would only be accepting "public" ids

Comment (achou11) [reply src/icon-server/fastify-plugin.js:13]: also updating the variable name to something like `PROJECT_PUBLIC_ID_REGEX`, maybe with the addition of a comment to summarize what's been explained here 😄

Comment (tomasciccola) [reply src/icon-server/fastify-plugin.js]: yeah, an also include every letter since its not hex (I think only lowercase, but left also upper just to be safe)

**Comment by gmaclennan** [comment]
not essential for this PR, but just a thought on how we expose "getProject". The Fastify "way" is to have a plugin that "decorates" the fastify instance with `fastify.getMapeoProject()`, and making that plugin the dependency of this and the blob http server. It's fine I think how it is implemented here, but if we start moving things into separate files if the code grows then it gets more complicated passing the getProject function to the enclosure (vs. attaching to context as a plugin).

File: src/icon-server/fastify-plugin.js
Comment (gmaclennan) [reply src/icon-server/fastify-plugin.js:51]: Should generally avoid creating instances within methods / requests - it has a performance impact.

**Review by gmaclennan — CHANGES_REQUESTED** [review]
Good work. Needs a couple of changes. Is the SQL changed here because of a change in schema in the stacked PR? I don't see any changes to schema in this PR.

File: src/icon-server/fastify-plugin.js
Comment (tomasciccola) [inline src/icon-server/fastify-plugin.js:45]: Type error here happens because I should somehow tell typescript that `fastify` param here is decorated with the `getProject` method. 
I tried a couple of things but couldn't make it work...
@gmaclennan ideas??

**Review by gmaclennan — CHANGES_REQUESTED** [review]
all looks good! good work. Hopefully my comment helps resolve the type issue, otherwise just @ts-ignore for now

**Comment by tomasciccola** [comment]
> all looks good! good work. Hopefully my comment helps resolve the type issue, otherwise just @ts-ignore for now

yeah, thanks for that. I saw that fastify documentation, just was trying to avoid using .d.ts and using jsdoc directly...

**Review by gmaclennan — APPROVED** [review]
Great job. Lgtm.

**Review by achou11 — APPROVED** [review]

File: src/mapeo-manager.js
Comment (achou11) [inline src/mapeo-manager.js:116]: didn't realize we were going with a decorator approach, which seems reasonable but maybe not completely necessary imo. with it, looks like we need to declare the dependency on this decorator whenever we register the plugins (https://fastify.dev/docs/latest/Reference/Decorators#decoratename-value-dependencies)

```ts
async function getProjectDecorator (fastify, opts) {
  this.#fastifyServer.decorate(
    'getProject',
    /** @param {string} projectId */
    async (projectId) => {
      return this.getProject(projectId)
    }
  )
}
```

```suggestion
    this.#fastifyServer.register(getProjectDecorator, 'getProject')

    this.#fastifyServer.register(IconServerPlugin, { prefix: '/icons/' }, ['getProject'])
```

File: types/fastify.d.ts
Comment (achou11) [inline types/fastify.d.ts:1]:
```suggestion
import { MapeoProject } from '../src/mapeo-project.js'
```

**Review by achou11 — COMMENTED** [review]
Couple of small comments on things I noticed (mostly for myself). Keeping in mind that with #365, some other changes are probably going to be needed here, but otherwise generally in a good state

File: src/mapeo-manager.js
Comment (achou11) [reply src/mapeo-manager.js:116]: hm after trying to implement for the blobs plugin, starting to question the benefit of the decorator, especially since you'd have to provide some kind of opt for defining how to get the project anyways. Maybe just keeping it to a `getProject()` opt on the icon/blob plugin is fine as opposed to the decorator?

**Comment by achou11** [comment]
heads up that i did a rebase after #335 was merged into main. entirely possible that I messed some stuff up but hopefully not too badly 😄

**Review by achou11 — COMMENTED** [review]
Still need to write a couple of unit tests specific to the plugin (e.g. plugin usage and url matching). Thinking about potentially holding off on e2e until this is merged and the media server work is done, but still on the fence about it

**Review by gmaclennan — APPROVED** [review]
Looks good to me. lgtm

**Comment by gmaclennan** [comment]
I merged main, since it was a lot of my sync changes on there so thought it easiest for me to resolve the merge conflicts

**Pull request merged**
